Pallet Actors: asynchronous smart contract design via message passing

sorpaas opened the pull request:
Note that everything's still heavily work-in-progress. The PR is open to document the design. cc @pepyakin 

### Overview

Actor-based smart contract model acts as an alternative to account-based smart contract model. Instead of direct calls, it uses message passing to invoke and interact with other smart contracts. Actors are easier to reason about and arguably make smart contracts safer to design, because a contract no longer need to care about invocations of other contracts. Another major benefit is that it provides much better backward compatibility guarantee -- the environment an actor interacts is limited, and as a result, upgrades accidentally breaking past deployed smart contracts will be much rarer, if not impossible.

### Messages

A message contains a payload together with a balance value. In other words, messages can send funds.

### Account types

Below are different account types as in the actor model. The main type is **actors**, with **libraries** and **system actors** supporting its operations. Another participants in this account model is **externally owned accounts**, which we'll briefly talk about later.

#### Actors

Actors can be thought as a function as below:

```
actor(message, account_data) -> (new_account_data, outgoing_messages, new_actor)
```

The wasm binary of an actor exposes two functions, `deploy(payload)` which is called when the contract is deployed, and `process(message)` which is a function that processes one message (with multiple messages forming loops of `process`). The contract only has information about its internal states, namely its storage data and balance. It interacts with the world state by sending and receiving messages.

The `process` loop of actors not called within transactions, but at the end of the block processing. The runtime decides the proper workload suitable for the current block, and process messages of actors (by calling `process(message)` of them). This allows the runtime to handle sudden spikes of incoming messages by reliving them over multiple blocks.

#### System actors

Note that actors do not have access to block information, nor can it create other contracts themselves. Instead, those operations are handled by system actors. Special actors, with the same interface of sending and receiving messages, handle contract creation, timestamp and randomness fetching, etc. The major benefit of designing it like this is that it allows the interface to have better backward compatibility -- when we update a system interface, simply create a new system actor at a different address, with no impact on existing contracts.

#### Libraries

Libraries are code blobs that can be imported by actors. They are mostly here to save state size, and is optional in the actor-based account model. Libraries do not have internal states.

#### Externally-owned accounts

EOA can send and receive messages to and from actors, as ways to invoke functions of contracts, and to receive funds.

#### Accessors

Accessors is an optional component in this actor-based account model. It can define functions that reads any world state and return values, but cannot modify any world state. Accessors are helper functions for end-users to access actor's internal data.

### Metering

Each actor has its own gas counter. The gas is used when the process loop is being executed, and once is exhausted, the runtime will not execute that actor until new funds have been added to process additional messages.

### Discussions

#### Message-passing consistency

Consistency of message-passing to the same actor is ensured, but not different ones. For example, in a process loop, if `A` emits messages `X`, `Y`, `Z` to `B`, `C` and then `B` again. We ensure that `X` is always processed before `Z`, but we make no promise of the execution order of `Y`.

#### Increased messages and invocation

This async design would require more message passing round compared with sync contract model. The impact (both to execution and storage) can be mostly limited by having a good schedule queue of process loop.

### Further readings

* https://corepaper.org/contract/actor/
* https://github.com/primea/design

NikVolf commented on frame/actors/src/exec.rs:
```diff
@@ -0,0 +1,35 @@
+	/// actor or if the message payload is too large.
+	fn send_message(&self, message: MessageFor<Self::T>) -> Result<(), &'static str>;
+	/// Get the message that the process function is currently operating on.
+	fn get_message(&self) -> MessageFor<Self::T>;
```
Should it be rather "dequeue", so it removes next message to process?

NikVolf replied:
Ah, nevermind!

I suppose `Ext` instance is what provided when handling particular message.

pepyakin commented:
First of all, thanks for the exploring of this direction. 

In the following I will attempt to dump all my thinking that I had so far on such a model that hopefully help in exploring this model.

## Further benefits of actor model

Actor is a bit like a blockchain in itself. Its state is determined by the "genesis" and the sequence of messages. Therefore:

1. Testing becomes a breeze. Just create a sequence of messages and test the resulting state.
1. It becomes more feasible to execute the actors in parallel. Every actor is executed in its own thread and the changes are independent from other actors.
1. If a client interested in a single actor, it can only subscribe to messages to it and apply them locally (provided the client also has data from the block, VM logic and the code of the actor). The client can have the full state of that actor if it likes to.

Furthermore,

1. It is easier to reason about the runtime side as well. There is no calling to the sandbox module within the supervisor (in case if wasm used as the VM).
1. Message passing usually means that there may be no observable difference between sending a message within the same chain or cross-chain. (This is why I think it is important to explore this direction: cross-chain/shard communication might become a common case)
1. Theoretically, actors are able to migrate cross-chain completely and transparently. (I mostly assume the SPREE-contracts case)

However, everything will depend on the final construction.

## Ultimate versioning - various VMs

I think of the backward compatibility a little bit broader in the sense that an actor can have vastly different virtual machines or state transition functions. As a rough example one actor could use wasm and another something entirely different VM or any STF really. In that sense, system actors would be actors with a fixed STF (fixed - that is, the behavior is not function of code of the actor) that are probably implemented within the runtime itself (I have opinion on the system contracts, about that later).

That said, I think that doesn't imply the guarantee regarding the gas costs, unfortunately. I can still imagine that some operations can get more expensive or cheaper. Ideally, we still make gas unobservable.

## System Actors

I am not sure that it is beneficial to make every interaction with the outer system through the system contracts. I think it might not be possible for some APIs and constructions. (For example, if we decide that not all messages are guaranteed to reach the recipients within the block, then the block info APIs (such as number) would be a mess.).
Furthermore, considering that we have more leeway given the black box nature of message processing I think system actors are not even necessary.

## Questions

### Dispatch of the messages

The questions around process loop should be addressed in the first place. For now, it is not clear for me what exactly happens with the messages sent by an actor during the processing of a message. How gas is topped up? I can see a problem that one actor can grief: a malicious actor can spam some innocent actor. Users/actors who are interested in using that actors must pay for these spam messages.

When I thought about this problem I came up with an idea that users must dispatch messages themselves paying for processing. The dispatched messages should follow rules of consistency, that is, the messages must be dispatched from a single source in order, but total from different source doesn't matter and in fact some sources can be ignored completely. However, I have no idea how to efficiently implement that.

### How to deal with the messages that induce a trap of the recipient? 

Our system here a bit departs from traditional actor systems, I believe. 

In traditional actor systems actors usually are restarted and messages are dropped. In our case, we don't have such a luxury. The state is persistent by default and the messages can have funds attached to them, so dropping a message is not a possibility.

### Are not accessors available to actors?

Just to confirm, if the accessors are available to call by the other actors. If they are available, I believe this would break the advantages I mentioned before: independent syncing, independent testing, location transparency, etc.

### Locality

Do we want to employ the principles of locality? That is, do we allow communications only thru `ActorRef` that were received by the actor, or created by it, or do we allow the actor to communicate with any actor whatsoever?

Security properties that this provides are tempting, however, it is not something that is not achievable without it. It is always possible to have ACL. And bookkeeping the references which the actor has access to has is additional overhead. So I am not sure if we actually want that.

### What is about logs/events?

LOGs / events have some similarity to messages, but not entirely. Namely, logs don't have a recipient and it doesn't make sense for them to carry funds.

sorpaas commented:
### Most APIs should be implemented as system actors

> I am not sure that it is beneficial to make every interaction with the outer system through the system contracts. I think it might not be possible for some APIs and constructions. (For example, if we decide that not all messages are guaranteed to reach the recipients within the block, then the block info APIs (such as number) would be a mess.).

If we want to make actors "pure functions" and accomplish your first stated goal that:

> Testing becomes a breeze. Just create a sequence of messages and test the resulting state.

Then in this case, those interactions must be through system actors. Thinking about designing APIs I think an important thing we have to decide is rather what the API users should *not* have access to. Actors can not be guaranteed to execute in a particular block, so things like block number may better just be hidden from users.

For actors that depend on functions such as scheduling or randomness, it is better to have dedicated system actors for them, rather than given them access to too-low-level functions and ask them to build it themselves.

The above describes the "strict mode" for actors, where information is only accessible if it's deterministic. This does provide some restrictions -- for example, an actor cannot access its own balance, but must use messages to track its balances by itself. This is because actor messages are not the only way a Substrate account may change its balance.

I think we definitely should support this "strict mode" for those blockchains that want full determinism. However, in the mean time, I think it's also okay if we add "relaxed mode" that adds those non-deterministic interfaces back for those blockchains that do not care about it too much.

### Messages and DoS attack

> I can see a problem that one actor can grief: a malicious actor can spam some innocent actor. Users/actors who are interested in using that actors must pay for these spam messages.

The sender will have to pay for an intrinsic fee for passing the message. So if we set the fee correctly for a chain I think DoS may not be easy. The attacker will have to burn a lot of money to DoS a single actor, and the runtime scheduler will also limit the effect of the attack.

The future queue is indeed hidden from actor's point of view. So scheduler can technically "re-order" them internally if it deems necessary, as long as it follows the consistency rules. In addition, we can also support contracts declaring static "filter rules" -- those is equivalent to `process(message)` function when a message is dropped immediately, but faster to be executed as it does not need to go through the VM.

### Deal with the messages that induce a trap of the recipient

> In traditional actor systems actors usually are restarted and messages are dropped. In our case, we don't have such a luxury. The state is persistent by default and the messages can have funds attached to them, so dropping a message is not a possibility.

In my view, a carefully designed actor should not trap on a message. An actor should always treat messages as un-trusted. If traps indeed happen, it indicate an internal logic error within the actor, and due to the deterministic nature of actor processing, the actor cannot be recovered unless there's social interventions. So that particular actor is just dead.

In another word, having money trapped in theDAO is better than having it stolen by hackers.

### Accessors and actors

> Are not accessors available to actors?

No I think accessors should not be available to actors. Accessors are mostly user-facing helper functions that deliver data to users. If an actor needs to fetch information from other actors, it should use the message passing interface in an RPC-like fashion, but not direct calls, as that will break the advantage of actor-based model. In some sense, accessors do not necessarily need to be on-chain.

In the mean time, we indeed can imagine having a "metadata system actor". This metadata system actor stores accessors, and reply to query messages from other actors by invoking the accessors.

### Locality

> Do we want to employ the principles of locality? That is, do we allow communications only thru ActorRef that were received by the actor, or created by it, or do we allow the actor to communicate with any actor whatsoever?

I have not really thought about this. I think I made the above design under the assumption that actors can communicate with any other actors. I see what you mean -- it's indeed quite tempting to have this model being locality/capatibility-based as it's *really really* convenient. I probably need to think more about this.

### Logs and events

> LOGs / events have some similarity to messages, but not entirely. Namely, logs don't have a recipient and it doesn't make sense for them to carry funds.

I think logs and events can just be directly added to ext functions. They can not have any effect (cannot modify/influence state in any way), and only emit data to off-chain observers.

pepyakin commented:
**System Actors**

I think these two approaches have the same degree of determinism (i.e. they are both fully deterministic) and actors in them both behave like pure functions.

The difference between them would be that the approach that involves only system actors won't require any additional context, only the actor STF and a string of messages will suffice for recreating the state of an actor.

The other approach, replacing the need in the system actors, that I had in mind is there are host functions that are able to return some state. To recreate the state of an actor the STF and the messages along with the context will be required.

There are different variations what exactly this state will encompass, but here is one possible option. Each message exists only within a block and that block provides the context. The context for an actor would give the information such as the block hash, number, the chain-specific data (like randomness, difficulty and what have you) and perhaps the actor's balance. This information should be read-only or can be precisely tracked by the actor independently. For example, we assume that the balance of an actor cannot be altered if the actor will get executed again within that block. Therefore, the balance of the actor can be precisely tracked independently by the VM given the balance at the start of the block and the sequence of messages to that actor. 

Moreover, as you mentioned that we assume that actors are special kinds of accounts and FRAME can theoretically alter their balances, this poses a TOCTOU problem for the system actors approach: the balance that the actor received might differ from the balance at the time of receiving and time of requesting if the message processing is spread across blocks.

Now, I think I sympathize the system actor approach because of its simplicity, however consider the following points:

1. Direct interactions via host functions has lower overhead, compared with involving the scheduler.
2. Direct interactions also would allow synchronous calls. 
3. A selective syncing (that is, syncing of a selected actor) would most likely require proving that all messages actually came from the claimed block, so it is not a big stretch to pass the other context data. Additionally, I think it is not unimaginable that a user might want to sync more than one actor and/or might want to relate the state of the actor with the state of the runtime. 
4. Testing would become a bit more annoying theoretically, since one will have to give the context even though none is used. But nothing unsurmountable.
5. And as I already mentioned, but worth reiterating, there is no difference regarding determinism, only the additional contextual data.
6. And finally, It is all or nothing. As soon as we need at least one context-dependent function we would automatically we would have a mix of interaction styles - system actors and host function. So we gotta be very careful here.

**State Bloat Managment**

The balance TOCTOU problem I mentioned above made me think that it would be really annoying if a state rent were employed since that requires pretty tight control of the balance from the state user side. The question is whether you think to go with the model where the actor pays for the state each block, has a proportional deposit, mix of the two (fwiw, this is what the pallet-contracts does), stateless approach or something else?

I expect actors to use storage more extensively since actors typically need to track the state of their execution. In synchronous contracts it often happens implicitly, i.e. just store something on the stack or heap make a sync call, update whatever you need. With actors this state must be persisted in the storage trie and probably loaded on the next message when the result has arrived.

**Gas Questions**

I think I wasn't clear enough, I meant that according to the original description all actors would have associated gas meter and I assume that everyone can top it up.

I imagine if I were a user of some actor, I'd need to top-up the gas meter of the actor I am interested in as well as other actors it might communicate with. If we are speaking about some isolated actor clump then if I overestimate what the actors consumes then other people can get a free ride or (be griefing) by essentially getting gas free. In a big actor constellation (think some big DeFi cluster) it becomes a huge mess.

Maybe I am missing something here, but that's precisely why I would like to hear the details.

Besides that, I think storing gas that carries over to the next block might have problems in that it loses it's ability to regulate demand on the block - buy low, spend high.

**Deal with the messages that induce a trap of the recipient**

Sure they not should not trap in the ideal world. And I admit that this is a simple and safe strategy. OTOH, it seems overly conservative for each and every case, especially if the contract is not responsible for a big sum of money (but, perhaps represents an agreement-like contract). Maybe we could leave this decision up to the contract itself?

**OOG condition semantics**

These points bring me to the question: how do you see handling an out-of-gas condition in the middle of message processing? It is really similar to a trap in a sense.

sorpaas commented:
Finished an updated draft about the design (https://corepaper.org/contract/actor/). Hopefully this addresses some issues @pepyakin raised before. Definitely not everything and I'm still thinking about some details (such as system actors).

A brief summary of what's changed:

* Instead of a contract address holding a single actor, now it holds an **actor group**. As its name implies, it's a collection of actors. Among them, there will be a main actor that handles messages from EOAs. Other than that main actor, other actors within this group cannot be called externally.
* There is now the concept of "child actor" or reversely "supervisor". Any actor can create new actors (of a specific `code_hash`) via externalities, and those new actors will be the direct child of the current actor. An actor can also `kill` its direct children.
* Message passing can now also pass actors. This moves ownership of an actor from one to another. Upon receiving, if it figures out that this is not an actor it wants, it can immediately kill it. This design is useful to enable patterns like "information fetcher".
* To actually create new contract addresses, an actor sends a message to the "address creation system actor" together with the new main actor encoded in the message.
* Added some notes about schedulers. The basic idea is that we can model it just like a normal process scheduler (like Linux's CFS), but just replace the concept of "time passed" with "gas consumed".

gnunicorn commented:
@sorpaas is this still relevant?

gnunicorn commented:
Closed due to it being stale for while.

sorpaas commented:
Sorry for keeping this stale for a while. There are two things I want to do before pushing more commits to the branch of this PR:

* Introduce the ability in Substrate for the runtime to save and load memory of sub-wasm instances. Right now we cannot do that -- we can create memory, but there's no function to inspect them. Being able to save and load memory will allow actor contracts to have a **single lifetime** for its entirety, and thus greatly simplify some of the contracts' design. For example, with this, we can store `Future` in memory persistently, allowing us to implement async future-based actor contracts and make them look almost the same with call-based routines.
* Abstract the wasm executive of `pallet-contracts` so that we can reuse it elsewhere (in particular, `pallet-actors`). This can remove the copy-paste code that currently exists in this PR.

sorpaas commented:
This is still work in progress, but I'm going to move the code of this PR into a separate repo. https://github.com/paritytech/artist/

I do want to try some significant changes of the design, which you can see the current document here. https://github.com/paritytech/artist/blob/master/docs/design.md

The thing is that I don't expect this to be merged in the short-term, and having PRs stale for a long time is not a good idea. It also brings some workflow issues.
